### DISCUSSION TITLE
Issues with container docker container build

### DISCUSSION THREAD

**Discussion by eggman920** [reflex-dev/reflex#1907] [General]
Would anyone be able to to help me understand what I'm doing wrong? I've tried many different things from the the docker documentation to referencing the following [https://github.com/tspycher/tspycher/tree/main](url) as well as #798 . I've added my current set up and I keep getting the following output when I run

`Info: Overriding config value api_url with env var API_URL=http://0.0.0.0:8000/
───────────────────────────── Starting Reflex App ──────────────────────────────
Detected database schema changes. Run reflex db makemigrations to generate 
migration scripts.
Warning: The path to the Node binary could not be found. Please ensure that Node
is properly installed and added to your system's PATH environment variable.
Info: Overriding config value api_url with env var API_URL=http://0.0.0.0:8000/`

before run I run any commands I run to ensure I'm in the project directory
`cd hundtry`
and then the following build commands
```
docker build -t reflex-app:latest . --build-arg API_URL=http://0.0.0.0:8000/

docker run -p 3000:3000 -p 8000:8000 --name app docker.io/library/reflex-app:latest
```
which results in the warning statement above.

I wasn't sure about naming so I tried following #798 and named some things the same, I'm also not sure about bun path, I've seen that may or may not be necessary because reflex already adds that for you

my current project structure is as follows:
```
hundtry
|--assests
|--hundtry
      |--hundtry.py
      |--__init__.py
|--requirements.txt
|--alembic
      |--versions
           |--78186c27bdca_.py
      |--script.py.mako
      |--env.py
|--alembic.ini
|--reflex.db
|--reconfig.py
|--Dockerfile
|--compose.yaml
|--Caddyfile
|--Caddy.Dockerfile
```

config:
```
import reflex as rx


config = rx.Config(
    app_name="hundtry",
    api_url="http://0.0.0.0:8000/",
    db_url="sqlite:///reflex.db",
    telemetry_enabled=False,
    )
```

Docker file:

```
# Stage 1: init
FROM python:3.11 as init

# Pass `--build-arg API_URL=http://app.example.com:8000/` during build
ARG API_URL

# Copy local context to `/app` inside container (see .dockerignore)
WORKDIR /app
COPY . .

# Create virtualenv which will be copied into final container
ENV VIRTUAL_ENV=/app/.venv
ENV PATH="$VIRTUAL_ENV/bin:$PATH"
RUN python3 -m venv $VIRTUAL_ENV

# Install app requirements and reflex inside virtualenv
RUN pip install -r requirements.txt


# Deploy templates and prepare app
RUN reflex init

# Export static copy of frontend to /app/.web/_static
RUN reflex export --frontend-only --no-zip

# Copy static files out of /app to save space in backend image
RUN mv .web/_static /tmp/_static
RUN rm -rf .web && mkdir .web
RUN mv /tmp/_static .web/_static

# Stage 2: copy artifacts into slim image 
FROM python:3.11-slim
ARG API_URL
WORKDIR /app
RUN adduser --disabled-password --home /app reflex
COPY --chown=reflex --from=init /app /app
USER reflex
ENV PATH="/app/.venv/bin:$PATH" API_URL=$API_URL

CMD reflex db migrate && reflex run --env prod --backend-only
```

below is the result of containter inspect command 
```

[
    {
        "Id": "sha256:7a7aabe6a2b8a0a179a3c8f5ad018ce4ebe6d0b7e5796e83c5bc3a91cc03a496",
        "RepoTags": [
            "hundtry:latest"
        ],
        "RepoDigests": [
            "hundtry@sha256:1d9faf9e9f8b0039c23a2dd83be34e41cd8aae6cb61e3a9fb975bfd6cd73fdee"
        ],
        "Comment": "buildkit.dockerfile.v0",
        "Created": "2023-10-02T18:00:57.057848468Z",
        "Author": "",
        "Config": {
            "User": "reflex",
            "AttachStdin": false,
            "Env": [
                "PATH=/app/.venv/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "LANG=C.UTF-8",
                "GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D",
                "PYTHON_VERSION=3.11.5",
                "PYTHON_PIP_VERSION=23.2.1",
                "PYTHON_SETUPTOOLS_VERSION=65.5.1",
                "PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/9af82b715db434abb94a0a6f3569f43e72157346/public/get-pip.py",
                "PYTHON_GET_PIP_SHA256=45a2bb8bf2bb5eff16fdd00faef6f29731831c7c59bd9fc2bf1f3bed511ff1fe",
                "API_URL=http://0.0.0.0:8000/"
            ],
            "Cmd": [
                "/bin/sh",
                "-c",
                "reflex db migrate \u0026\u0026 reflex run --env prod --backend-only"
            ],
            "WorkingDir": "/app"
        },
        "Architecture": "arm64",
        "Os": "linux",
        "Size": 454533120,
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:311627f8702d2d97e63b916cad711aa9f962a1f20da5572d68b8f223ff051e73",
                "sha256:c4d9cc0a506397e0ad36ddbdf72d19e613c14398db1885c15b5cdaaf3ad8a843",
                "sha256:157a5fb7cdeb684a5ef63681e3324581cf56d888ab940c9de34a542fe944d65c",
                "sha256:02436e886926662de870dd9bd0c59dc815918a5051df5270c5570f26722ccf21",
                "sha256:96c02c33d5bbc277e38c37e7eae43ec64c1a5bf1f8633d9112f5848769dbbb9d",
                "sha256:0ea3c64729940b0b9be30ec8ce753accbe27f0b82d2e481adbb59cf9a13efbe1",
                "sha256:a479687e67a60a7ce58419956f64a030b6ba1e6c82fd0c1dbca2c094134ff412",
                "sha256:0ed4b1c020a3df2d47e342abfc88dd9fb800edd0d5e2e27549dacf10c226f153"
            ]
        },
        "Metadata": {
            "LastTagTime": "0001-01-01T00:00:00Z"
        }
    }
]
```

**Comment by masenf** [comment]
Have you ever ran `reflex db makemigrations` in the project? Based on your root layout, it looks like your `alembic/versions` directory is empty... but if you have any schema defined, then it needs to have a migration script created to initialize it inside the container.

Second, side point, but `--build-arg API_URL=http://0.0.0.0:8000/` doesn't look right. The client (browser) needs to be able to resolve the API_URL to access the backend, so this should be something like your public IP address or hostname.

**Reply by eggman920** [reply]
I have ran `reflex db makemigrations` , the versions directory does contain a .py file the project (I will edit the original post to include that). Are you saying I need to add the `reflex db makemigrations`command before `reflex db migrate` in the final line of the docker file?

Also for the api_url do you mean something like http://localhost:8000 instead of 0.0.0.0:8000?

**Reply by masenf** [reply]
> Are you saying I need to add the `reflex db makemigrations` command before `reflex db migrate` in the final line of the docker file?

No, you you should only ever need `reflex db migrate` in the Dockerfile. `makemigrations` should always be run outside of the container (i.e. during development, when the schema changes).

> `Detected database schema changes`

This line in the output indicates that reflex noticed `rx.Model` metadata in your app that differs from what was created by apply `reflex db migrate`. 

What is the contents of your alembic migration script? And does it match your app's schema?

**Reply by chimeracle** [reply]
Here's the docker file that works for me. I set up a venv beforehand and it gets copied in. Some things didn't work for me with the python slim or buster versions, so I used the bigger base image. I'm probably doing something inefficient or wrong. Make sure your rxconfig file has your api_url set to something other than localhost or 0.0.0.0 that is externally accessible. Hope it'll point you in the right direction.

```
FROM python:3.9 as base_image

RUN set
ENV root_path /

# PRerequisites packages for nodejs, bun
RUN apt-get update
RUN apt-get install curl unzip --yes

# Nodejs current stable
RUN apt-get update && apt-get install -y nodejs npm
RUN curl -fsSL https://deb.nodesource.com/setup_16.x | bash - && apt-get install -y nodejs
RUN node -v
RUN python --version

# Python deps
WORKDIR /opt/app
COPY ./requirements.txt .
RUN pip install -r requirements.txt

# Cleaning
RUN apt-get autoremove --yes

# Copy codebase
COPY . .
RUN rm -f requirements.txt
RUN rm -f Dockerfile

RUN reflex init

# Set interface-
EXPOSE 3000
EXPOSE 8000

# default run
CMD ["reflex", "run", "--env", "prod"]
```
